File: usth/MATH2.2/labwork/1/ratpoison.m
function [x fx ea i] = ratpoison (f, df, x0, es = 0.00000001, imax = 20)
  nargin < 2 && error ('ratpoison requires at least 2 ingredients');
  [x fx dfx ea i] = deal (x0, f (x0), df (x0), 1, 0);
  while (ea > es && i++ < imax)
    [xold x] = deal (x, x - fx/dfx);
    [fx dfx] = deal (f (x), df (x));
    if (fx == 0)
      ea = 0;
      break;
    elseif (x)
      % just drop the percent BS
      ea = abs ((x - xold) / x);
    end
  end
end

File: usth/MATH2.2/labwork/1/ratpoison.py
def ratpoison(f, df, x, es=10**-8, imax=20):
    ea, i = 1, 0
    while ea > es and i < imax:
        i += 1
        xold, x = x, x - f(x)/df(x)
        if f(x) == 0: return x, 0, 0, i
        if x: ea = abs((x - xold) / x)
    return x, f(x), ea, i
